x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -490,7 +492,87 @@ static XContentBuilder buildUpdatedDocument(
+            return false;
+        }
+
+        final Map<String, Object> currentCreator = apiKeyDoc.creator;
+        final var user = authentication.getEffectiveSubject().getUser();
+        final var sourceRealm = authentication.getEffectiveSubject().getRealm();
+        if (false == (Objects.equals(user.principal(), currentCreator.get("principal"))
+            && Objects.equals(user.fullName(), currentCreator.get("full_name"))
+            && Objects.equals(user.email(), currentCreator.get("email"))
+            && Objects.equals(user.metadata(), currentCreator.get("metadata"))
+            && Objects.equals(sourceRealm.getName(), currentCreator.get("realm"))
+            && Objects.equals(sourceRealm.getType(), currentCreator.get("realm_type")))) {
+            return false;
+        }
+        if (sourceRealm.getDomain() != null) {
+            if (currentCreator.get("realm_domain") == null) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            final var currentRealmDomain = RealmDomain.fromXContent(
+                XContentHelper.mapToXContentParser(
+                    XContentParserConfiguration.EMPTY,
+                    (Map<String, Object>) currentCreator.get("realm_domain")
+                )
+            );
+            if (sourceRealm.getDomain().equals(currentRealmDomain) == false) {
+                return false;
+            }
+        } else {
+            if (currentCreator.get("realm_domain") != null) {
+                return false;
+            }
+        }
+
+        final Map<String, Object> newMetadata = request.getMetadata();
+        if (newMetadata != null) {
+            if (apiKeyDoc.metadataFlattened == null) {
+                return false;
+            }
+            final Map<String, Object> currentMetadata = XContentHelper.convertToMap(apiKeyDoc.metadataFlattened, false, XContentType.JSON)
+                .v2();
+            if (newMetadata.equals(currentMetadata) == false) {
+                return false;
+            }
+        }
+
+        final List<RoleDescriptor> newRoleDescriptors = request.getRoleDescriptors();
+        if (newRoleDescriptors != null) {
+            final List<RoleDescriptor> currentRoleDescriptors = parseRoleDescriptorsBytes(
+                request.getId(),
+                apiKeyDoc.roleDescriptorsBytes,
+                RoleReference.ApiKeyRoleType.ASSIGNED
+            );
+            if (false == (newRoleDescriptors.size() == currentRoleDescriptors.size()
+                && new HashSet<>(newRoleDescriptors).equals(new HashSet<>(currentRoleDescriptors)))) {
+                return false;
+            }
+        }
+
+        assert userRoles != null;
+        final List<RoleDescriptor> currentLimitedByRoleDescriptorRoles = parseRoleDescriptorsBytes(
+            request.getId(),
+            apiKeyDoc.limitedByRoleDescriptorsBytes,
+            RoleReference.ApiKeyRoleType.LIMITED_BY
```

Comment:
This might be OK though the behaviour is not perfect.

The method `parseRoleDescriptorsBytes` replaces old superuser role definition (7.x) with the new one (8.0). This means it can return a role descriptor that is different from what is actually stored. If we only test against the returned role descriptor, we may miss an update because the equality does not apply to the actual stored descriptor. 

In practice this can only happen in a mix version cluster (8.4 + 7.17). If update skips for a `superuser` limitedByRoleDescriptor. It may not be too big of a deal especially since the legacy descriptor will be handled correctly when loaded (at authentication time). 

Alternatively, we can check whether the current doc has version 7.x and the returned a descriptor is a `superuser`. If answers to both are `true`, it is _not_ a noop (because we will be updating the legacy superuser descriptor to the new one). This would also means after update, the key will lose the legacy superuser behaviour even when it is used on a 7.17 node. But that is probably OK (and even desired due to consistency). 

It is definitely an edge case. I don't really have a strong opinion. But at least it worths some code level comment. What do you think?